### PR TITLE
Correctly transform spreads to use proper concat method

### LINKED ISSUE #8907: @babel/plugin-transform-spread regression of spread array
# v7 Regression

> First check out: https://new.babeljs.io/docs/en/next/v7-migration.html
> Also a partial upgrade tool: https://github.com/babel/babel-upgrade

**Potential Commit/PR that introduced the regression**
If you have time to investigate, what PR/date introduced this issue.

https://github.com/babel/babel/pull/6763

before that PR `[...array]` was transpiled to `[].concat(array)` now it's  `array.concat()`, this change breaks immutable array(seamless-immutable), for example

```
import Immutable from `seamless-immutable`
const array = Immutable([1, 2, 3]);
console.log(Immutable.isImmutable([...array]);
```
it prints `false` in console of latest Chrome and in Node v8.11.3, but it prints true with Babel@7, it should be `false`, but as it's transpiled to `array.concat()`, and an immutable array returns immutable array too with builtin methods(`concat`, `map`, `reduce`...)

**Describe the regression**
A clear and concise description of what the regression is.

**Input Code**

```js
var your => (code) => here;
```

**Babel Configuration (.babelrc, package.json, cli command)**

```js
{
  "your": { "config": "here" }
}
```

**Expected behavior/code**
A clear and concise description of what you expected to happen (or code).

**Environment**
- Babel version(s): [e.g. v6.0.0, v7.0.0-beta.34]
- Node/npm version: [e.g. Node 8/npm 5]
- OS: [e.g. OSX 10.13.4, Windows 10]
- Monorepo [e.g. yes/no/Lerna]
- How you are using Babel: [e.g. `cli`, `register`, `loader`]

### PR DESCRIPTION
**Pull request by danez**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #8907 
| Patch: Bug Fix?          | y
| Major: Breaking Change?  | n
| Minor: New Feature?      | n
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This reverts some of the optimizations of #6763.

Transforming a spread to this

```js
// [...arr, b]
arr.concat(b)
```
is dangerous as we do not know what `arr` is exactly. It might be a subclass of `Array` with changed behavior of its concat function.

Instead I changed it to use

```js
Array.prototype.concat(arr, b);
```

Other option would be to use what was there before #6763

```js
[].concat(arr, b);
```

which is smaller, but needs to create an additional array.

Also added comments to hopefully make the logic understandable.

### CODE REVIEW

File: packages/babel-plugin-transform-destructuring/test/fixtures/destructuring/array-unpack-optimisation/output.js [hunk 1]
```diff
@@ -21,7 +21,7 @@ var _ref4 = [a[1], a[0]];
 a[0] = _ref4[0];
 a[1] = _ref4[1];
 
-var _ref5 = babelHelpers.toConsumableArray(foo).concat([bar]),
+var _ref5 = Array.prototype.concat(babelHelpers.toConsumableArray(foo), [bar]),
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-destructuring/test/fixtures/destructuring/array-unpack-optimisation/output.js]: I'm not sure I love the idea of relying on the implicit `Array.prototype` as the `this` context for the `concat` call.

Comment (danez) [reply packages/babel-plugin-transform-destructuring/test/fixtures/destructuring/array-unpack-optimisation/output.js]: Should i change it to `[].concat(..)` ?

Or i guess we could do 
```
var _a = babelHelpers.toConsumableArray(foo);
Array.prototype.concat.call(_a, _a, ..);
```

Comment (danez) [reply packages/babel-plugin-transform-destructuring/test/fixtures/destructuring/array-unpack-optimisation/output.js]: Okay it is now transforming to `[].concat()` again.

**Comment by smashercosmo** [comment]
One bad thing about this PR is that before it if somebody tries to spread Set in loose mode, clear error is thrown, but now he will get wrong output of ```[Set()]```, which will be harder to spot.

**Comment by rubennorte** [comment]
I agree with @smashercosmo. This also allows to spread non-iterable values again, which is wrong (e.g.: `[...null]`). I know this is for loose mode but IMO having a subset of the semantics (only accepting arrays) is better than changing the semantics.

**Comment by fdanielsen** [comment]
Sorry if this is the wrong place to comment, but this has references to a lot of related history.

In a modern browser that supports the spread operator natively, `[...el.querySelectorAll('.foo')]` correctly converts the iterable `NodeList` to an array of elements. But `[].concat(el.querySelectorAll('.foo')` gives an array of a single item which is the `NodeList`.

This doesn't seem like the intended result with Babels support of the spread operator?

**Comment by calebeby** [comment]
@fdanielsen I'm seeing the same thing as you, with `DOMTokenList` as well (from "spreading" `element.classList`

**Comment by jridgewell** [comment]
The loose transform only handles Arrays. You'll have to use the spec transform (as well as polyfilling `Symbol.iterator` methods for the iterable) if you want this to be done correctly.

**Comment by calebeby** [comment]
I don't see a `spec` option here: https://babeljs.io/docs/en/babel-plugin-transform-spread

Are the docs out of date? Happy to send a PR if that is the case 😃

**Comment by jridgewell** [comment]
Spec in this case is the default. We only add an explicit `spec` option if the doing it by default would cause a large performance regression or something else undesirable.

**Comment by calebeby** [comment]
Got it. Thanks for the clarification 👍

**Pull request merged**